Backend/core/comms.py

```diff
@@ -0,0 +1,181 @@
+import json, select, socket, struct, sys, time
+import threading
+import traceback
+import aiohttp
+import asyncio
+import config
+
+from . import db
+
+format_string = '<' # little-endian
+byte_length = 0
+properties = []
+frontend_data = {}
+solar_car_connection = {'lte': False, 'tcp': False}
+# Convert dataformat to format string for struct conversion
+# Docs: https://docs.python.org/3/library/struct.html
+types = {'bool': '?', 'float': 'f', 'char': 'c', 'uint8': 'B', 'uint16': 'H', 'uint64': 'Q'}
+
+def set_format(file_path: str):
+    global format_string, byte_length, properties
+    with open(file_path, 'r') as f:
+        data_format = json.load(f)
+
+    for key in data_format.keys():
+        format_string += types[data_format[key][1]]
+        byte_length += data_format[key][0]
+        properties.append(key)
+        config.FORMAT[key] = {'type': data_format[key][1]}
+
+def unpack_data(data):
+    fields = {}
+    unpacked_data = struct.unpack(format_string, data)
+    for i in range(len(properties)):
+        fields[properties[i]] = unpacked_data[i]
+    return fields
+
+
+class Telemetry:
+    __tmp_data = {'tcp': b'', 'lte': b''}
+    latest_tstamp = 0
+
+    def listen_tcp(self, server_addr: str, port: int):
+        print(f'connecting to {server_addr}:{port}')
+        global solar_car_connection, frontend_data
+        while True:
+            # create a client socket
+            client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            client.settimeout(5)
+            try:
+                client.connect((server_addr, port))
+                solar_car_connection['tcp'] = True
+                print('connected')
+            except Exception:
+                time.sleep(3)   # macOS is the superior system, it teaches you an important lesson of having patience
```

Comment:
Yeah, this is a temporary fix, I believe it has something to do with the macOS implementation taking up too much thread time and blocking other threads from executing. we should definitely look into better practices for this, I'll make this one an issue as well.

Reply:
... actually, since we are switching to udp soon and since udp is connectionless, it should eliminate thread time on retrying. We'll see how udp preforms and decide if we need performance improvement.